enhancements/ignition-spec-dual-support.md

```diff
@@ -0,0 +1,327 @@
+Terraform provider that seem to be incompatible with this change.
+
+** MCO details: **
+
+A spec translator will first be implemented in the MCO, with the ability to
+detect untranslatable configs. The MCO then should be updated to have support
+for both ignition V0S2 and V2S3. Since the MCO is responsible for the current
+cluster nodes, it will be the only place at which spec translation is done.
+The translation will happen when the version of MCO with dual support and
+translator is first deployed; it will detect the existing config being spec 2,
+generate a new renderedconfig based on a translator from spec2 to spec 3, and
+do a “dummy reboot” (if necessary) to bring the cluster to the desired config,
+now in spec 3. If the translation fails, the MCO will throw an alert to the
+admin that the update has failed with reason and (if untranslatable configs
+were detected) which lines of the existing config were not translatable.
+
+The spec translator will also translate the existing bootimage configs served
+to new nodes joining the cluster. At this point the MCS will be able to serve
+two boot configs of the same effect but different ignition spec versions. The
+MCS must also be updated to detect, upon a new image joining the cluster, what
+version the ignition inside that system is on. Then it will serve the correct
+config, and apply the other machineconfigs as normal. A spec 2 bootimage will
+also go through an extra step (like before) of first having the spec on the
+node translated to spec 3.
+
+The MCO should also add functionality to more easily reconcile broken
+machineconfigs, and allow the user to correctly recover/abort a failed update
+to spec 3.
+
+
+** Other notes: **
+
+Spec 2 -> Spec 3 translation has not been fully implemented anywhere before.
+There could be many edge cases we have not yet considered. There are other
+potential difficulties such as serving the correct ignition config. See above
+section on risks and mitigations.
+
+Starting from some version of Openshift, likely
+v4.${release-with-dual-support + 1} we can remove dual support and be fully
+ignition spec 3. This is because upgrade is supported for version x -> x+1
+-> x+2 and not x -> x+2 directly.
+
+Kubernetes 1.16 onwards has support for CRD versioning:
+https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/.
+If we opt to delay this that is potentially an alternative method of
+implementation.
+
+
+### Test Plan
+
+Extensive testing of all possible paths, especially those outlined in the
+user stories, is critical to the success of this major update. The existing
+CI infrastructure is a good start for upgrade testing. There should be
+additional tests added, especially in the MCO repo, for edge cases as
+described in the user stories, to ensure we never break this behaviour.
+Many existing tests will also have to be updated given the spec change.
+
+
+### Upgrade / Downgrade Strategy
```

Comment:
Downgrades from spec 3 to spec 2 are off the table, right?

Reply:
Yes. We should never have to do that. We do however probably need rollback functionality to recover from bad machineconfig states.